src/main/java/net/revelc/code/impsort/ImpSort.java

```diff
@@ -166,4 +178,55 @@ private static void convertAndAddImport(LinkedHashSet<Import> allImports, List<N
     allImports.add(imp);
   }
 
+  /*
+   * Extract all of the tokens from the main body of the file.
+   *
+   * This set of tokens represents all of the file's dependencies, and is used
+   * to figure out whether or not an import is unused.
+   */
+  private static Set<String> extractTokens(CompilationUnit unit) {
+    Set<String> tokens = new LinkedHashSet<String>();
+    NodeList<TypeDeclaration<?>> types = unit.getTypes();
+    types.forEach((Node node) -> {
+      Optional<TokenRange> tokenRange = node.getTokenRange();
+      if (tokenRange.isPresent() && tokenRange.get() != TokenRange.INVALID) {
+        JavaToken token = tokenRange.get().getBegin();
+        while (token != null && token != JavaToken.INVALID) {
+          String tokenString = token.toString();
+          if (!tokenString.isEmpty() && Character.isLetter(tokenString.charAt(0))) {
+            tokens.add(token.toString());
+          }
+          Optional<JavaToken> next = token.getNextToken();
+          if (next.isPresent()) {
+            token = next.get();
+          } else {
+            break;
+          }
+        }
+      }
+    });
+
+    return tokens;
+  }
+
+  /*
+   * Remove unused imports.
+   *
+   * This algorithm only looks at the file itself, and evaluates whether or not
+   * a given import is unused, by checking if the last segment of the import path
+   * (typically a class name or a static function name) appears in the file.
+   *
+   * This means that it is not possible to remove import statements with wildcards.
+   */
+  private static void removeUnusedImports(Set<Import> imports, Set<String> tokens) {
+    imports.removeIf(i -> {
+      String[] segments = i.getImport().split("\\.");
+      if (segments.length == 0) return false;
```

Comment:
Wouldn't it be nice to handle invalid code graciously as well? I can certainly remove this check if you really want me to, but it just feels like good defensive programming to leave it in.

Reply:
Perhaps it's fine, but it looks like dead code to me, since it can never happen. Invalid code which would cause a compilation failure would fail long before this stage, within the parser itself. However, it's possible a newer version of the parser will allow partial processing of invalid code.  To protect against that case, instead of returning false, I would suggest `throw new AssertionError("Parse tree includes invalid import statements")`